### PR TITLE
2271 feedback reward custom metrics

### PR DESCRIPTION
**Pull request by axsaucedo**
PR includes:
* Extension to support custom metrics in feedback
* Extra tags to make metrics logged per request unique
* Simple example using statistical metrics for binary classification

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **axsaucedo**
You can assign the PR to them by writing `/assign @axsaucedo` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Mon Aug 17 13:25:39 UTC 2020
The logs for [lint] [2] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2289/2.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2289 --build=2

**Comment by seldondev** [comment]
Mon Aug 17 13:25:58 UTC 2020
The logs for [pr-build] [1] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2289/1.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2289 --build=1

**Comment by seldondev** [comment]
Mon Aug 17 16:10:56 UTC 2020
The logs for [pr-build] [3] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2289/3.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2289 --build=3

**Comment by seldondev** [comment]
Mon Aug 17 16:11:12 UTC 2020
The logs for [lint] [4] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2289/4.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2289 --build=4

**Comment by seldondev** [comment]
@axsaucedo: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | e5b96265f0031a572e58b77cadf3272e8695669d | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@axsaucedo: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | e5b96265f0031a572e58b77cadf3272e8695669d | [link]() | `/test this`
lint | e5b96265f0031a572e58b77cadf3272e8695669d | [link]() | `/test lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Mon Aug 17 18:57:12 UTC 2020
The logs for [pr-build] [5] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2289/5.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2289 --build=5

**Comment by axsaucedo** [comment]
Rebasing master

**Comment by seldondev** [comment]
Mon Aug 17 19:00:43 UTC 2020
The logs for [integration] [9] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2289/9.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2289 --build=9

**Comment by seldondev** [comment]
Mon Aug 17 19:00:43 UTC 2020
The logs for [pr-build] [8] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2289/8.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2289 --build=8

**Comment by seldondev** [comment]
Mon Aug 17 19:01:00 UTC 2020
The logs for [lint] [10] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2289/10.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2289 --build=10

**Comment by seldondev** [comment]
@axsaucedo: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | e5b96265f0031a572e58b77cadf3272e8695669d | [link]() | `/test this`
lint | 5f265ef4f57cfccf3093934c4e7de2f5851bf054 | [link]() | `/test lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@axsaucedo: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint | 5f265ef4f57cfccf3093934c4e7de2f5851bf054 | [link]() | `/test lint`
pr-build | 5f265ef4f57cfccf3093934c4e7de2f5851bf054 | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by axsaucedo** [comment]
Integration tests pass, and python tests pass locally, not sure why lint and pr-build are failing. Lint seems to be failing due to licenses, but pr-build seems to be failing on the operator tests which has nothing to do with pr...
image

/test this
/test lint

**Comment by seldondev** [comment]
Tue Aug 18 05:28:27 UTC 2020
The logs for [lint] [12] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2289/12.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2289 --build=12

**Comment by seldondev** [comment]
Tue Aug 18 05:28:39 UTC 2020
The logs for [pr-build] [11] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2289/11.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2289 --build=11

**Comment by seldondev** [comment]
@axsaucedo: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | 5f265ef4f57cfccf3093934c4e7de2f5851bf054 | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by axsaucedo** [comment]
It seems lint test is now fixed and PR build tests are passing except for a flaky test 
/test this

**Comment by seldondev** [comment]
Tue Aug 18 06:00:35 UTC 2020
The logs for [pr-build] [13] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2289/13.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2289 --build=13

**Comment by axsaucedo** [comment]
Seems it's good to go 
image

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *[jklaise](https://github.com/SeldonIO/seldon-core/pull/2289#pullrequestreview-469439072 "Approved")*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **axsaucedo**
You can assign the PR to them by writing `/assign @axsaucedo` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by axsaucedo** [comment]
Thanks @jklaise! In regards to your first question, the answer is definitely, the next PR will be to add documentation to provide insights on the existing capabilities but also the current discussions (as the existing endpoints will be depricated). 

In regards to the second question, that is a very good point, I could've used the `feedback` function in this situation. In the other example I had to send the ID so it would not have been possible to use the `feedback` function.  Atm the feedback function requires providing a proto message for the request/response so it's a bit of an overhead complexity, but we'll probably want to re-think the python client function as well to make it simpler to use than sending a full request

File: python/seldon_core/metrics.py
Comment (RafalSkolasinski) [inline python/seldon_core/metrics.py:83]: We could probably later on change this list of constants into one `enum`

**Review by RafalSkolasinski — APPROVED** [review]
Nice one!

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[jklaise](https://github.com/SeldonIO/seldon-core/pull/2289#pullrequestreview-469439072 "Approved")*, *[RafalSkolasinski](https://github.com/SeldonIO/seldon-core/pull/2289#pullrequestreview-470254024 "Approved")*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [RafalSkolasinski,jklaise]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Wed Aug 19 09:08:40 UTC 2020
The logs for [pr-build] [14] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2289/14.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2289 --build=14

**Comment by seldondev** [comment]
Wed Aug 19 09:08:54 UTC 2020
The logs for [lint] [15] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2289/15.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2289 --build=15

**Comment by seldondev** [comment]
@axsaucedo: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint | 5f265ef4f57cfccf3093934c4e7de2f5851bf054 | [link]() | `/test lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
New changes are detected. LGTM label has been removed.

**Comment by seldondev** [comment]
Wed Aug 19 13:32:47 UTC 2020
The logs for [pr-build] [16] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2289/16.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2289 --build=16

**Comment by seldondev** [comment]
Wed Aug 19 13:34:09 UTC 2020
The logs for [lint] [17] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2289/17.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2289 --build=17

**Comment by seldondev** [comment]
@axsaucedo: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint | 6e1e790131271d29ad013a38974b388adc61c6ed | [link]() | `/test lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Pull request merged**
